client/main-layout.tsx

```diff
@@ -0,0 +1,401 @@
+  const dispatch = useAppDispatch()
+  const inGameplayActivity = useAppSelector(s => s.gameplayActivity.inGameplayActivity)
+  const isEmailVerified = useAppSelector(s => s.auth.user.emailVerified)
+  const isMatchmakingSearching = useAppSelector(s => !!s.matchmaking.searchInfo)
+  const lobbyCount = useAppSelector(s => s.serverStatus.lobbyCount)
+  const starcraft = useAppSelector(s => s.starcraft)
+
+  const [searchingMatchOverlayOpen, setSearchingMatchOverlayOpen] = useState(false)
+
+  const findMatchButtonRef = useRef<HTMLButtonElement>()
+  const searchingMatchButtonRef = useRef<HTMLButtonElement>()
+
+  useEffect(() => {
+    dispatch(openChangelogIfNecessary())
+
+    if (firstLoggedIn) {
+      firstLoggedIn = false
+      if (!isEmailVerified) {
+        dispatch(
+          addLocalNotification({
+            id: EMAIL_VERIFICATION_ID,
+            type: NotificationType.EmailVerification,
+          }),
+        )
+      }
+
+      dispatch(addPrivacyPolicyNotificationIfNeeded())
+      dispatch(addTermsOfServiceNotificationIfNeeded())
+      dispatch(addAcceptableUseNotificationIfNeeded())
+    }
+
+    return () => {
+      firstLoggedIn = true
+    }
+  }, [isEmailVerified, dispatch])
+
+  const onFindMatchClick = () => {
+    if (!isShieldBatteryHealthy({ starcraft })) {
+      dispatch(openDialog(DialogType.ShieldBatteryHealth))
+    } else if (!isStarcraftHealthy({ starcraft })) {
+      dispatch(openDialog(DialogType.StarcraftHealth))
+    } else {
+      // TODO(2Pac): Remove `any` once the `openOverlay` is TS-ified
+      dispatch(openOverlay('findMatch') as any)
+    }
+  }
+
+  const onCreateLobbyClick = () => {
+    if (!isShieldBatteryHealthy({ starcraft })) {
+      dispatch(openDialog(DialogType.ShieldBatteryHealth))
+    } else if (!isStarcraftHealthy({ starcraft })) {
+      dispatch(openDialog(DialogType.StarcraftHealth))
+    } else {
+      // TODO(2Pac): Remove `any` once the `openOverlay` is TS-ified
+      dispatch(openOverlay('createLobby') as any)
+    }
+  }
+
+  const onJoinLobbyClick = () => {
+    if (!isShieldBatteryHealthy({ starcraft })) {
+      dispatch(openDialog(DialogType.ShieldBatteryHealth))
+    } else if (!isStarcraftHealthy({ starcraft })) {
+      dispatch(openDialog(DialogType.StarcraftHealth))
+    } else {
+      // TODO(2Pac): Remove `any` once the `openOverlay` is TS-ified
+      dispatch(openOverlay('joinLobby') as any)
+    }
+  }
+
+  const onMapUpload = (map: MapInfoJson) => {
+    // TODO(2Pac): Remove `any` once the `openOverlay` is TS-ified
+    dispatch(openOverlay('browseServerMaps', { ...serverMapsProps, uploadedMap: map }) as any)
+  }
+
+  const onMapDetails = (map: MapInfoJson) => {
+    dispatch(openDialog(DialogType.MapDetails, { mapId: map.id }))
+  }
+
+  const onRemoveMap = (map: MapInfoJson) => {
+    dispatch(removeMap(map))
+  }
+
+  const onRegenMapImage = (map: MapInfoJson) => {
+    dispatch(regenMapImage(map))
+  }
+
+  const serverMapsProps = {
```

Comment:
This is a circular dependency, and once you useCallback onUploadMap that becomes very clear:

onUploadMap uses serverMapProps which uses onUpoadMap

It's also declared *after* it's being used, which is something you should generally avoid. I don't think this really needs to exist as an object and you should probably pass these props directly instead of doing stuff like this (this is also bad from a typing perspective because it is effectively `any`)